power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/dataset.hpp

```diff
@@ -259,14 +438,23 @@ template <dataset_type_tag dataset_type_> class Dataset {
         ComponentInfo const& info = dataset_info_.component_info[component_idx];
         Buffer const& buffer = buffers_[component_idx];
         auto const ptr = reinterpret_cast<StructType*>(buffer.data);
-        if (scenario < 0) {
-            return std::span<StructType>{ptr, ptr + info.total_elements};
-        }
-        if (info.elements_per_scenario < 0) {
-            return std::span<StructType>{ptr + buffer.indptr[scenario], ptr + buffer.indptr[scenario + 1]};
+        auto const total_range = std::span<StructType>{ptr, ptr + info.total_elements};
+        return get_span_impl<std::span<StructType>>(buffer, info, scenario, total_range);
+    }
+
+    // get non-empty columnar buffer
+    template <class StructType>
+    RangeObject<StructType> get_columnar_buffer_span_impl(Idx scenario, Idx component_idx) const {
+        // return empty span if the component does not exist
+        if (component_idx < 0) {
+            return {};
         }
-        return std::span<StructType>{ptr + info.elements_per_scenario * scenario,
-                                     ptr + info.elements_per_scenario * (scenario + 1)};
+        // return span based on uniform or non-uniform buffer
+        ComponentInfo const& info = dataset_info_.component_info[component_idx];
+        Buffer const& buffer = buffers_[component_idx];
+        assert(is_columnar(buffer));
+        RangeObject<StructType> const total_range{info.total_elements, buffer.attributes};
+        return get_span_impl<RangeObject<StructType>>(buffer, info, scenario, total_range);
```

Comment:
optional nitpick: use template argument deduction

```suggestion
        return get_span_impl(buffer, info, scenario, total_range, RangeObject<StructType>{info.total_elements, buffer.attributes});
```